packages/eslint-plugin/src/rules/no-implied-eval.ts

```diff
@@ -156,7 +156,11 @@ export default util.createRule({
       }
 
       const [handler] = node.arguments;
-      if (EVAL_LIKE_METHODS.has(calleeName) && !isFunction(handler)) {
+      if (
+        EVAL_LIKE_METHODS.has(calleeName) &&
+        !isFunction(handler) &&
+        !context.getScope().variables.some(v => v.name === calleeName)
```

Comment:
this isn't *quite* correct!

You'll find that this won't work for cases like this:
```ts
function setTimeout() {}

{
  setTimeout();
}
```
because the declaration for `setTimeout` is in the parent scope.

Additionally - a variable can and will be created from the "globals" config - so if you're parsing in "script" mode, then `setTimeout` will  be in the global scope and it'll break.
We parse our tests in "module" mode though - it's rare that you actually want script mode now-a-days.

We need to:
1) get the definition
2) ensure the definition IS an "implicit global variable"

Reply:
Interesting @bradzacher, do you have a test case that fails? https://github.com/typescript-eslint/typescript-eslint/pull/4049/commits/52a2b3193c07fb635551ec40afca1bc4e010df3e has a working one.

Reply:
sorry - I was being lazy cos mobile.
`setTimeout()` is ignored by the rule because it has no arguments.
The correct test case would be
```ts
function setTimeout() {}

{
  setTimeout(100);
}
```

Scope analysis is actually kind of an ugly API in this regard unfortunately.
What you end up with is a scope tree that looks like this:
```
+ Scope#1 - Global Scope
+---> Variables = [Variable#1 (setTimeout (implicit global))]
+---> References = []
|
+--+ Scope#2 - Module Scope
   +---> Variables = [Variable#2 (setTimeout (local declaration)]
   +---> References = []
   |
   +--+ Scope#3 - Block Scope
      +---> Variables = []
      +---> References = [Reference#1 (Variable#2)]
```

So when you do `context.getScope()` you get back `Scope#3`. Then you inspect the variables and you find no variables at all (so your `.some` returns `false`).

So there are two ways you can do this:
1) crawl up the scope tree until you find a scope with a variable declared with the name, then check if it's a "global" variable.
2) follow the reference from the current scope to its variable and check if it's a "global" variable.

```
function isReferenceToGlobalFunction(calleeName: string): boolean {
  const ref = context.getScope()
    .references.find(ref => ref.identifier.name === calleeName);
  if (ref == null || ref.resolved == null) {
    return true;
  }
  
  const variable = ref.resolved;
  // ensure it's the "global" version
  return variable.defs.length === 0;
}
```

Reply:
Fascinating, thanks for the explanation!